### PR TITLE
Add cooldown for unregistered namechanges

### PR DESCRIPTION
**Pull request by AnnikaCodes**
If it would be more useful with a different cooldown length (I just grabbed 2 minutes off the top of my head) or enforcing a cooldown on namechanges to registered alts as well, I'm happy to change it.

### CODE REVIEW

**Comment by Zarel** [comment]
If it's not done yet, the first step is to make semilocked/locked user namechanges not show up in the log at all.

**Comment by Zarel** [comment]
Probably also modchatted and muted users, tool

**Comment by Zarel** [comment]
I don't know how I feel about a throttle. Maybe just combine successive renames, like we do joins/leaves. Or not show renames at all, for users who haven't talked recently?

**Comment by scheibo** [comment]
Agree theres probably a couple things we want here to cut down on spam. Combining alone still seems like it may be annoying? Unless you collapse all the renames into 1? Agree to not showing renames at all for people not talking 'recently' (whats the constant there?), but some sort of throttling mechanism beyond that still seems desirable - these are done by bots who could pretty easily break collapsing i think by just also interspersing text in between their name changes. Throttling is heavy handed, but what we're trying to target here is malicious spam bots

**Comment by Zarel** [comment]
> Unless you collapse all the renames into 1?

Yes, that's what I meant.

> Throttling is heavy handed, but what we're trying to target here is malicious spam bots

A 1 per 2 minute throttle just feels like it could affect some legitimate use-cases (like actual typoes), which is why I'm trying to think of alternate solutions. If it was like 3 per 2 minutes, I'd probably feel better.

**Comment by AnnikaCodes** [comment]
> locked



> If it's not done yet, I think an important first step is to make semilocked/locked user namechanges not show up in the log at all.

This is currently implemented, but only when modchat is set — I added a fix so that it works regardless.

> Yes, that's what I meant. If there's no chat between "A renamed to B" and "B renamed to C", collapse that into "A renamed to C".

This is a client-side problem — see https://github.com/smogon/pokemon-showdown-client/pull/1565.

> A 1 per 2 minute throttle just feels like it could affect some legitimate use-cases (like actual typoes), which is why I'm trying to think of alternate solutions. If it was like 3 per 2 minutes, I'd probably feel better.

Sure.

File: server/rooms.ts [hunk 1]
```diff
@@ -592,11 +592,8 @@ export abstract class BasicRoom {
 		this.destroy();
 	}
 	reportJoin(type: 'j' | 'l' | 'n', entry: string, user: User) {
-		let reportJoins = this.reportJoins;
-		if (reportJoins && this.settings.modchat && !user.authAtLeast(this.settings.modchat, this)) {
-			reportJoins = false;
-		}
-		if (reportJoins) {
+		const canTalk = user.authAtLeast(this.settings.modchat ?? 'unlocked', this) && !this.isMuted(user);
+		if (this.reportJoins && canTalk) {
```
Comment (Zarel) [inline server/rooms.ts:596]: Oh, this code is a lot better.

Does this cover semilocks?

Comment (Zarel) [reply server/rooms.ts:596]: Looks like no, can you add that condition? Maybe with an exception for players (i.e. any roomauth)

File: server/users.ts [hunk 2]
```diff
@@ -709,6 +715,23 @@ export class User extends Chat.MessageContext {
 		this.s2 = tokenDataSplit[6];
 		this.s3 = tokenDataSplit[7];
 
+		if (!this.trusted && userType === '1') { // userType '1' means unregistered
+			const elapsed = Date.now() - this.lastNewNameTime;
+			if (elapsed < NAMECHANGE_THROTTLE) {
+				if (this.newNames >= NAMES_PER_THROTTLE) {
+					this.send(
+						`|nametaken|${name}|You must wait ${Math.round((NAMECHANGE_THROTTLE - elapsed) / 1000)} more
```
Comment (mia-pi-git) [inline server/users.ts:723]: ``
```suggestion
						`|nametaken|${name}|You must wait ${Chat.toDurationString(NAMECHANGE_THROTTLE - elapsed)} more
```

File: server/rooms.ts (hunk 1 above)
Comment (AnnikaCodes) [reply server/rooms.ts:596]:
```ts
authAtLeast(minAuth: string, room: BasicRoom | null = null) {
		if (!minAuth || minAuth === ' ') return true;
		if (this.locked || this.semilocked) return false;
...
```
`User#authAtLeast` will return false if a user is semilocked. I can still add an exception for players / voices in the battle, if you would like.

File: server/users.ts (hunk 2 above)
Comment (Zarel) [reply server/users.ts:723]: Oh, good suggestion!

File: server/rooms.ts (hunk 1 above)
Comment (Zarel) [reply server/rooms.ts:596]: Ohhh, yeah, that works.

I do think semilocked players should be able to talk in their own battles, yes.

Comment (Zarel) [reply server/rooms.ts:596]: They wouldn't be able to rename spam anyway, because after the first rename they're no longer players.

**Comment by Zarel** [comment]
A good improvement as-is, in any case.

**Pull request merged**